### PR TITLE
A capture that records the driver's own conversation (#62)

### LINKED ISSUE #62: Capture mode that records the active driver's own conversation
## What is missing

The raw-bus capture is **passive by design**: `rs485Task` runs it *instead of* its poll cycle, and the bridge transmits nothing for the whole window. That is the right shape for the problem it was built for — an unidentified device, where mixing our own traffic into the recording would present two conversations as one stream (`capture_runner.h` says so explicitly).

It also means the capture **cannot record a protocol we already speak**, because the driver that would produce that traffic is the thing being paused.

## Confirmed on hardware, 2026-07-26

Run against the production bridge and its EverSolar TL3000-20, 9600/none/8/1, 30 s:

```json
{"status":"done","elapsed_ms":30002,
 "summary":{"frames":0,"bytes":0,"modbus_crc_ok":0,"aa55_frames_ok":0}}
```

Zero frames, exactly as the design implies: the EverSolar is a slave on the AA55 protocol, the bridge is the only master on that bus, and the bridge was silent for thirty seconds.

The run was not wasted — it validated everything else about the path on production hardware: the bus lock was taken and released, `restoreDriverLine()` put the line back, polling resumed on the next cycle (755 → 756), the device never went stale or offline, and `idle_gap_ms: 4` confirmed the t3.5 derivation at 9600. **What it could not do is the one thing it was wanted for.**

## Proposal: a second mode, explicitly named

A capture that runs **alongside** normal polling and records the driver's own request/response pairs. Not a replacement for the passive mode — a sibling, chosen at request time.

Value:
- Verify a shipped driver's framing against its protocol document (`docs/eversolar-protocol.md`) on real hardware, which today can only be done with a USB-RS485 tap and a laptop at the inverter.
- Turn a "works for me / not for me" bug report into a byte-level artefact, for drivers where the reporter has no tap.
- Regression evidence after a protocol change, from the field rather than from fixtures.

## Design questions

1. **How is our own traffic distinguished from the device's?** With the bridge transmitting, direction matters and the recording is no longer one anonymous stream. Marking each frame TX or RX is the obvious answer, but is that knowable from the transport, or does the driver have to say so?
2. **Does it still cut on idle gaps?** With request/response structure available, framing could follow the transaction instead of t3.5 — which would also lift the 19200 ceiling documented for the passive mode.
3. **Does it need the bus lock at all?** The passive mode holds it for the whole window. This one must not, or it defeats its own purpose.
4. **Should the passive mode's card in the UI mention this one?** Today the capture card only appears when discovery names nothing (`if(!c.length)`), so on a bridge with a working driver there is no way to reach a capture from the UI at all — you have to POST to `/api/v1/actions/capture` by hand. Whatever this becomes probably belongs somewhere reachable.

## Out of scope

Decoding. This produces bytes and framing metadata; interpretation stays with `tools/decode_eversolar.py` and its successors.

### PR DESCRIPTION
**Pull request by Timdebruijn**
Closes #62.

## The problem, in one measurement

The raw bus capture is **passive by design**: `rs485Task` runs it *instead of* a poll, so the bridge is silent for the whole window. That is right for an unidentified device. It also means it cannot record a protocol we already speak — the driver that would produce that traffic is the thing it pauses. Thirty seconds against a live inverter with a working driver: `frames: 0, bytes: 0`.

## The observation that decided the design

The issue left two questions open. One observation answers both:

- both protocol layers hand a **complete request to a single `write()` call** ([pmu_transaction.cpp:27](src/protocols/pmu/pmu_transaction.cpp:27), [modbus_client.cpp:37](src/protocols/modbus/modbus_client.cpp:37))
- every driver reaches the bus through `Transport`, and `Rs485Transport` is the only transport compiled into the firmware

So **direction is knowable at the transport** — no driver is asked, no driver changed, and every driver added later is covered the day it arrives. And the record boundary can follow the transaction instead of guessing at silence.

That lifts the passive mode's documented ceiling. It cuts on the t3.5 idle gap, honest at 9600 and 19200 and approximate above, because at 38400 the real gap is finer than any reader resolves. Here the primary boundary is the direction turning around, exact at any baud rate. The gap still cuts when the direction does not change, and **every record says which rule made it** — a boundary the protocol made and one the recorder imposed are otherwise indistinguishable, and only the first kind means "these bytes are one frame".

## It arms; it does not take the bus

The branch in `rs485Task` deliberately does **not** `continue`. Every other branch there takes the iteration away from polling because it needs the bus; taking it here would reproduce exactly the bug this fixes.

Three failure classes the passive mode has to handle do not exist: no bus lock, no line reconfiguration, and therefore no `lineReconfigured` / `restoreDriverLine()` dance. The line is read off the bridge rather than supplied — there is a working driver, so it is a fact.

Cost in the RS485 hot path when nothing is recording: **one inline null check**.

## API — additive

`mode` defaults to `passive`, so every existing caller keeps its meaning and `GET /api/v1/capture` is byte-for-byte what it was.

**Deviation from the plan, stated rather than done quietly.** The plan had one GET returning either report with a `mode` field. Two independent runners keep two independent last-reports, so one route would need new state to decide which to return, and a consumer would branch on `mode` anyway — with half the fields absent, which reads as missing data rather than as a different kind of report. So `GET /api/v1/capture/driver` is its own route.

The four line parameters are **rejected with 400** in `mode=driver`, not ignored: accepting a baud rate and recording at a different one would put a number in the report the capture never ran at.

## What it can tell you that the passive mode cannot

- **`sent` against `received`.** Requests with no replies means the bridge is talking and the device is not answering. The passive mode cannot reach that diagnosis — there the bridge is the silent one, so a dead bus and an unanswering device produce the same empty report.
- **The gap on a reply is the device's response time** — the number a driver's read timeout has to cover, and the one to look at when an inverter answers by hand and intermittently in service.
- **A shipped driver can be checked against its protocol document** on real hardware, with no USB-RS485 tap and laptop at the inverter.

## UI

The existing capture card lives under `if(!c.length)` — it only appears when discovery names nothing, which is exactly the bridge this mode cannot help. The driver capture goes on **Health**, reachable whenever the bridge is working.

Verified in a browser against `tools/preview_web.py`, not assumed from a passing syntax check: the card renders, a four-frame report comes out as `→` / `←` with `turn` / `silence` / `split`, and both failure verdicts read as intended.

## Things I changed my mind about while building

- **Wrote an `abandon()` and deleted it.** Nothing that would take the bus can start while this is armed, so it had no caller. An unused escape hatch invites a future caller to reach for it instead of asking why the guard exists.
- **`service()` returns the transition, not a bool.** A bool leaves "armed" and "still recording" indistinguishable, which is how a once-per-capture log line becomes one per iteration.
- **Discovery is refused during a driver capture**, for a reason the other two do not share: it would not conflict at all — it would happily record straight through a probe sweep — and the report would then show something no driver ever does.

## Verification

- **953 native tests**, 26 new. Six mutations, each aimed at one decision: reversing the direction/gap check order, letting an empty read count as a turn-around, removing the direction cut, removing the window guard, taking a `TransportLock` in `service()`, swapping the tx/rx labels. All caught.
- One mutation found a **missing test rather than a bug**: an empty read counting as a turn-around only misbehaves when it follows a *request*, which no test covered — and every driver's read loop does exactly that on its first fruitless read.
- One assertion of mine asserted nothing (`0 == 0` under a standing clock) and was replaced.
- `check_layering.sh` all ten rules — rule 1 caught two brand names in my own header comment.
- `check_web_js.py`, `check_dashboard_layout.py`, `check_measurement_ids.py`, `ruff`: clean.
- Firmware builds with no warnings from our own sources. RAM 22.5%, flash 22.9%.

## Not verified

**Nothing here has run on hardware yet.** Per the agreed gate: merge on host tests and CI, then OTA to the production bridge and compare poll duration against the known baseline — ewma 103 ms, max 111 ms, min 97 ms. A recorder in the hot path that moves those measurably is a finding, not a detail. The real acceptance test is a driver capture against the live inverter, held next to `docs/eversolar-protocol.md`.

### CODE REVIEW

File: src/diagnostics/bus_tap.cpp [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+    lastByteMs_ = nowMs;
+
+    // Closed HERE rather than left open for finish() to deal with. Left open, the record
+    // survives to the end of the window and finish() labels it WindowEnd -- which is a false
+    // account when the budget ran out with most of the window still to go, and cut_by exists
+    // precisely so a reader does not have to guess at that. Every later feed() is refused by
+    // the cap check above, so nothing would have been added to it anyway.
+    if (ranOutOfRoom) {
+        closeFrame(CutReason::CaptureFull);
+    }
```
[the remark this answers is not in the document]
Comment (Timdebruijn) [reply src/diagnostics/bus_tap.cpp]: Right, and reproduced before fixing: an 8-byte budget filled by 4 + 4 gave `truncated=0`, `cutBy=WindowEnd`. My fix flagged the case inside the byte loop, so it only fired when a write went *past* the budget — a write whose last byte lands exactly on it leaves the loop normally.

The condition is now asked of the state *after* the loop rather than flagged inside it: "the budget is gone" is the whole condition, however the loop happened to leave. Test added for the exact-fill boundary, and mutation-proved by restoring the flag-inside-the-loop.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: tools/check_layering.sh [hunk 2]
```diff
@@ -246,6 +246,31 @@ tracked_dirs = {str(pathlib.PurePosixPath(t).parent) for t in tracked}
...
+    s = re.sub(r"[^\w\s-]", "", s)
+    return re.sub(r"\s", "-", s)
+
+
+def anchors_in(path: pathlib.Path) -> set:
+    return {
+        slug(line)
+        for line in path.read_text(encoding="utf-8").splitlines()
+        if line.startswith("#")
+    }
```
[the remark this answers is not in the document]
Comment (Timdebruijn) [reply tools/check_layering.sh]: Correct, and worse than it sounds: I counted 13 phantom anchors across two documents, several of them from the capture examples added in this very branch (`# Heliograph driver capture ...`). That makes the rule fail *open* — a link to a heading that no longer exists could match a phantom and be waved through, which is worse than no check because the check is trusted.

Fenced blocks are now skipped. Proved by pointing a link at one of the phantoms and watching the fixed rule reject what the previous version accepted.

Worth noting for anyone editing it: writing the literal fence marker in that heredoc breaks the *shell* parser — it sits inside a `$(...)` substitution, where bash still tracks backticks as legacy command substitution even with a quoted delimiter. The marker is built with `chr(96)` and the comment says why.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_
